internal/ingress/controller/controller.go

```diff
@@ -559,9 +561,15 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
 			}
 		}
 
+		// set aside canary ingresses to merge later
 		if anns.Canary.Enabled {
-			glog.Infof("Canary ingress %v detected. Finding eligible backends to merge into.", ing.Name)
-			mergeAlternativeBackends(ing, upstreams, servers)
+			canaryIngresses = append(canaryIngresses, ing)
+		}
+	}
+
+	if safeToMergeCanaryIngresses(ingresses, canaryIngresses) {
+		for _, canaryIng := range canaryIngresses {
+			mergeAlternativeBackends(canaryIng, upstreams, servers)
```

Comment:
Since this is done before we set https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#default-backend, canary feature won't work when the main backend has not endpoints and Nginx is proxying to custom default backend. IMO this is an expected behaviour. Not the concern of this PR but it would  be nice to have an e2e test case where main ingress's backend has not endpoints (replicas: 0) and assert that all requests return 503 regardless of canary cookie/header being set to `always`.

Reply:
That's a good point. I'll add an e2e for this